library/core/src/slice/mod.rs

```diff
@@ -3607,6 +3615,200 @@ impl<T> [T] {
+    #[unstable(feature = "slice_align_to_ish", issue = "none")]
+    #[inline]
+    pub fn aligned_subslice_mut(&mut self, align: usize) -> &mut [T] {
+        let size = crate::mem::size_of::<T>();
+        assert!(
+            size < align,
+            "aligned_subslice does nothing for alignments below or at the element type's alignment"
+        );
+        assert!(
+            align % size == 0,
+            "aligned_subslice only works for alignments that are multiples of the element's size"
+        );
+        let offset = self.as_ptr().addr() % align;
+        // SAFETY: See the `align_to_mut` method for the detailed safety comment.
+        let end_offset = unsafe { self.as_ptr().offset(self.len() as isize) }.addr() % align;
+        let end = self.len() - (align / size - end_offset);
+        &mut self[offset..end]
+    }
+
+    /// Transmute the slice elements to another type.
+    ///
+    /// If the target element type is smaller than the source element type, the
+    /// returned slice will have multiple elements per element of the original slice.
+    ///
+    /// Cannot be used to go to an element type with higher alignment requirements.
+    /// Use `aligned_subslice` for that instead.
+    ///
+    /// # Panics
+    ///
+    /// The element sizes and the slice length must be such that all elements of the source
+    /// slice fit exactly into a slice of the destination element type. Resize your input slice
+    /// before invoking `transmute_elements` to uphold this checked requirement.
+    ///
+    /// # Safety
+    ///
+    /// This method is essentially a `transmute` between different elements, and even from
+    /// multiple elements into a single one or vice versa, so all the usual caveats
+    /// pertaining to `transmute::<T, U>` also apply here.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// unsafe {
+    ///     let ints: [u32; 2] = [1, 2];
+    ///     let smaller_ints = ints.transmute_elements::<u16>();
+    ///     assert_eq!(smaller_ints.len(), 4);
+    /// }
+    /// ```
+    #[must_use]
+    #[unstable(feature = "slice_align_to_ish", issue = "none")]
+    #[track_caller]
+    pub const unsafe fn transmute_elements<U>(&self) -> &[U] {
+        const {
+            let align_u = crate::mem::align_of::<U>();
+            let align_t = crate::mem::align_of::<T>();
+            assert!(align_u <= align_t, "use `aligned_subslice` instead");
+        };
+        let size_u = crate::mem::size_of::<U>();
+        let size_t = crate::mem::size_of::<T>();
+        if size_u > size_t {
+            assert!(
+                self.len() * size_u % size_t == 0,
+                "input slice does not fit exactly into a slice of the output element type"
+            );
+        } else {
+            assert!(
+                self.len() * size_t % size_u == 0,
+                "input slice does not fit exactly into a slice of the output element type"
+            );
+        }
```

Comment:
Do the panics get eliminated when the types are exact multiples in size and alignment?